firmware/src/Adafruit_BME680_rpi.cpp

```diff
@@ -0,0 +1,624 @@
+//                         void *interface);
+static int8_t spi_read(uint8_t reg_addr, uint8_t *reg_data, uint32_t len,
+                       void *interface);
+static int8_t spi_write(uint8_t reg_addr, const uint8_t *reg_data, uint32_t len,
+                        void *interface);
+static void delay_usec(uint32_t us, void *intf_ptr);
+
+unsigned long millis();
+
+void delay(uint32_t ms);
+
+
+
+/*!
+ *  @brief  Reads 8 bit values over SPI
+ */
+static int8_t spi_read(uint8_t reg_addr, uint8_t *reg_data, uint32_t len,
+                       void *intf_ptr) {
+  MySPI *_dev = (MySPI *)intf_ptr;
+
+  reg_addr |= 0x80;
+
+  _dev->read(reg_addr,reg_data,len);
+
+  return 0;
+
+  // if (!_dev->write_then_read(&reg_addr, 1, reg_data, len, 0x0)) {
+  //   return -1;
+  // }
+
+  // return 0;
+}
+
+/*!
+ *  @brief  Writes 8 bit values over SPI
+ */
+static int8_t spi_write(uint8_t reg_addr, const uint8_t *reg_data, uint32_t len,
+                        void *intf_ptr) {
+  MySPI *_dev = (MySPI *)intf_ptr;
+
+  _dev->write(reg_addr,reg_data,len);
+
+  // if (!_dev->write((uint8_t *)reg_data, len, &reg_addr, 1)) {
+  //   return -1;
+  // }
+
+  return 0;
+}
+
+// Shim functions so copied Adafruit code can use arduino names
+
+static void delay_usec(uint32_t us, void *intf_ptr) {
+  (void)intf_ptr; // Unused parameter
+  sleep_us(us);
+}
+
+
+unsigned long millis()
+{
+  return to_ms_since_boot(get_absolute_time());
+}
+
+void delay(uint32_t ms)
+{
+  sleep_ms(ms);
+}
+
+
+// PUBLIC FUNCTIONS
+
+Adafruit_BME680::Adafruit_BME680(spi_inst_t* spi_hw,
+           uint8_t spi_pico_pin, uint8_t spi_poci_pin, uint8_t spi_sck_pin,
+           uint8_t cs_pin, bool init_spi_hardware)
+{
+
+  _spidev = new MySPI(spi_hw,spi_pico_pin,spi_poci_pin,spi_sck_pin,cs_pin,init_spi_hardware);
+  
+}
+
+
+/*!
+ *  @brief  Initializes the sensor
+ *          Hardware ss initialized, verifies it is in the I2C or SPI bus, then
+ * reads calibration data in preparation for sensor reads.
+ *  @param  addr
+ *          Optional parameter for the I2C address of BME680. Default is 0x77
+ *  @param  initSettings
+ *          Optional parameter for initializing the sensor settings.
+ *          Default is true.
+ *  @return True on sensor initialization success. False on failure.
+ */
+bool Adafruit_BME680::begin(uint8_t addr, bool initSettings) {
+  int8_t rslt;
+  
+  gas_sensor.chip_id = 0;
+  gas_sensor.intf = BME68X_SPI_INTF;
+  gas_sensor.intf_ptr = (void *)_spidev;
+  gas_sensor.read = &spi_read;
+  gas_sensor.write = &spi_write;
+
+  gas_sensor.amb_temp = 25; /* The ambient temperature in deg C is used for
+                               defining the heater temperature */
+  gas_sensor.delay_us = delay_usec;
+
+
+  rslt = bme68x_init(&gas_sensor);
+#ifdef BME680_DEBUG2
+  printf("Init Result: ");
+  printf("%i", rslt);
+#endif
+
+  if (rslt != BME68X_OK)
+    return false;
+
+#ifdef BME680_DEBUG
```

Comment:
With "Serial" created in Arduino.cpp and Arduino.h that we're adding, this unaltered section from the original should "just work."